File: app/src/main/java/com/ivanovsky/passnotes/data/repository/keepass/KeepassNoteRepository.java
package com.ivanovsky.passnotes.data.repository.keepass;

import androidx.annotation.NonNull;

import com.ivanovsky.passnotes.data.entity.OperationResult;
import com.ivanovsky.passnotes.data.repository.NoteRepository;
import com.ivanovsky.passnotes.data.repository.encdb.dao.NoteDao;
import com.ivanovsky.passnotes.data.entity.Note;

import java.util.List;
import java.util.UUID;

public class KeepassNoteRepository implements NoteRepository {

    private final NoteDao dao;

    KeepassNoteRepository(NoteDao dao) {
        this.dao = dao;
    }

    @NonNull
    @Override
    public OperationResult<List<Note>> getNotesByGroupUid(UUID groupUid) {
        return dao.getNotesByGroupUid(groupUid);
    }

    @NonNull
    @Override
    public OperationResult<Integer> getNoteCountByGroupUid(UUID groupUid) {
        OperationResult<Integer> result = new OperationResult<>();

        OperationResult<List<Note>> getNotesResult = dao.getNotesByGroupUid(groupUid);
        if (getNotesResult.getObj() != null) {
            result.setObj(getNotesResult.getObj().size());
        } else {
            result.setError(getNotesResult.getError());
        }

        return result;
    }

    @NonNull
    @Override
    public OperationResult<UUID> insert(Note note) {
        return dao.insert(note);
    }

    @NonNull
    @Override
    public OperationResult<Note> getNoteByUid(UUID uid) {
        return dao.getNoteByUid(uid);
    }

    @NonNull
    @Override
    public OperationResult<UUID> update(Note note) {
        return dao.update(note);
    }

    @NonNull
    @Override
    public OperationResult<Boolean> remove(UUID noteUid) {
        return dao.remove(noteUid);
    }
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/data/repository/keepass/KeepassNoteRepository.kt
package com.ivanovsky.passnotes.data.repository.keepass

import com.ivanovsky.passnotes.data.entity.Note
import com.ivanovsky.passnotes.data.repository.encdb.dao.NoteDao
import com.ivanovsky.passnotes.data.repository.NoteRepository
import com.ivanovsky.passnotes.data.entity.OperationResult
import com.ivanovsky.passnotes.data.entity.Property
import java.util.UUID

class KeepassNoteRepository(private val dao: NoteDao) : NoteRepository {

    override fun getNotesByGroupUid(groupUid: UUID): OperationResult<List<Note>> {
        return dao.getNotesByGroupUid(groupUid)
    }

    override fun insert(note: Note): OperationResult<UUID> {
        return dao.insert(note)
    }

    override fun getNoteByUid(uid: UUID): OperationResult<Note> {
        return dao.getNoteByUid(uid)
    }

    override fun update(note: Note): OperationResult<UUID> {
        return dao.update(note)
    }

    override fun remove(noteUid: UUID): OperationResult<Boolean> {
        return dao.remove(noteUid)
    }

    override fun find(query: String): OperationResult<List<Note>> {
        val getNotesResult = dao.all
        if (getNotesResult.isFailed) {
            return getNotesResult.takeError()
        }

        val allNotes = getNotesResult.obj
        val matchedNotes = allNotes.filter { note -> note.matches(query) }

        return OperationResult.success(matchedNotes)
    }

    private fun Note.matches(query: String): Boolean {
        return title.contains(query, ignoreCase = true) ||
            properties.any { property -> property.matches(query) }
    }

    private fun Property.matches(query: String): Boolean {
        return name?.contains(query, ignoreCase = true) == true ||
            value?.contains(query, ignoreCase = true) == true
    }
}